lib/Sema/DerivedConformanceDifferentiable.cpp

```diff
@@ -851,19 +853,36 @@ static void checkAndDiagnoseImplicitNoDerivative(TypeChecker &TC,
     if (!vd->hasInterfaceType())
       continue;
     auto varType = DC->mapTypeIntoContext(vd->getValueInterfaceType());
-    if (vd->getAttrs().hasAttribute<NoDerivativeAttr>() ||
+    if (vd->getAttrs().hasAttribute<NoDerivativeAttr>())
+      continue;
+    // Check whether to diagnose stored property.
+    bool conformsToDifferentiable =
         TC.conformsToProtocol(varType, diffableProto, nominal,
-                              ConformanceCheckFlags::Used))
+                              ConformanceCheckFlags::Used).hasValue();
+    bool isConstantProperty = vd->isLet() && vd->hasInitialValue();
+    // If stored property should not be diagnosed, continue.
+    if (conformsToDifferentiable && !isConstantProperty)
       continue;
+    // Otherwise, add an implicit `@noDerivative` attribute.
     nominal->getAttrs().add(
         new (TC.Context) NoDerivativeAttr(/*Implicit*/ true));
     auto loc =
         vd->getLoc().isValid() ? vd->getLoc() : DC->getAsDecl()->getLoc();
     assert(loc.isValid() && "Expected valid source location");
-    TC.diagnose(loc, diag::differentiable_implicit_noderivative_fixit,
-                vd->getName())
+    // Diagnose stored property with fixit.
+    if (!conformsToDifferentiable) {
+      TC.diagnose(loc, diag::differentiable_implicit_noderivative_fixit,
+                  vd->getName())
+          .fixItInsert(vd->getAttributeInsertionLoc(/*forModifier*/ false),
+                       "@noDerivative ");
+      continue;
+    }
+    TC.diagnose(
+          loc,
```

Comment:
Slightly awkward indentation.

Reply:
I hate it too. The diagnostic name is too long, this is what's suggested by `clang-format`.